Added support for FKeys when switching gametabs

honeyhoney opened the pull request:
Requires the default keybinding setup.
FKey:Boolean defaults to false. It's a completely optional functionality.

slackydev commented:
I would like to have something changed before I merge this.
A tab isn't usually opened directly through `Gametabs`, but through eg `Inventory.Open`, so what I suggest is a fieldvar in the gametabs structure that's used to toggle on/off using function keys. Instead of as a argument to `.Open`.

``` pascal
TRSGametab = record
  Tabs: TBoxArray;
  Keybindings: TIntegerArray;
  FKeysEnabled: Boolean;
end;

procedure TRSGametab.ToggleFKeys(Enable:Boolean); 
...
```

This would as you see work globally, so that we always open tabs in a specific way.

And then ...

``` pascal
function TRSGametab.Open(Tab:EGametab; tries:Int32=2): Boolean;
...
  if (FKeysEnabled and (Tab <> tabLogout)) then
    Keyboard.PressKey(111+keybindings[Ord(tab)])
  else
    Mouse.Click(B, 1);
```

When you mentioned this yesterday I also thought it would be nice with mouse as a fallback, and to disable using function keys if it fails once. As some scripts may enable this, while some users of that script are not aware of this causing the script to not work for everyone. And as I also mentioned Simba on Linux does not support sending function keys input iirc.

honeyhoney commented:
Good point. I hadn't considered the fact that TRSGametab.Open isn't always going to be called directly.
I did consider using the mouse as a falback but didn't bother with it.

I'll get on with the changes you've suggested and let you take another look.

honeyhoney commented:
Updated. It's worth noting that if the FKey fails to open the gametab it will then attempt to use the mouse. However, the FKey usage does not count as a "try" at that point in time. It is only when the mouse is used (and then fails) that a try is counted. For subsequent attempts (tries) the FKey usage is disabled so it will only use the mouse from that point on.

slackydev commented:
This looks good, I see a couple of small things that I would like to see changed, after that I think we are set! :+1: 

``` pascal
Keybindings := [VK_F1, VK_F2, VK_F3, VK_ESCAPE, VK_F4, VK_F5, VK_F6, VK_F7,
                VK_F8, VK_F9, 0, VK_F10, VK_F11, VK_F12];
```

and shorten `.Open` into this (untested) code (test first please):

``` pascal
function TRSGametab.Open(Tab:EGametab; tries:Int32=2): Boolean;
var
  i:Int32;
  B:TBox;
begin
  B := self.GetBounds(Tab);
  if Self.IsOpen(Tab) then
    Exit(True);

  for i:=1 to tries do
  begin
    if (Self.FKeysEnabled and (Keybindings[Ord(tab)] <> 0)) then
    begin
      Keyboard.PressKey(Keybindings[Ord(tab)]);
      if self.IsOpen(Tab, Random(2500,3500)) then
        Exit(True);
      Self.ToggleFKeys(False);
    end;

    Mouse.Click(B, 1);
    if self.IsOpen(Tab, Random(2500,3500)) then
      Exit(True);
  end;
end; 
```

This makes it possible for users to set their own keybindings.

honeyhoney commented:
Updated with suggested changes. Tested with correct and incorrect keybindings (including attempting to enable FKeys again).

(also squashed the commits so hopefully it'll look cleaner if this gets merged in)

TheTS commented:
U gettin' fancy gurl

slackydev commented:
Merged, thanks! :D
